hawkular-datamining-forecast/src/main/java/org/hawkular/datamining/forecast/AutomaticForecaster.java

```diff
@@ -193,4 +219,103 @@ private TimeSeriesModel selectBestModel(List<DataPoint> dataPoints) {
 
         return dataPoints;
     }
+
+    /**
+     * Strategy used for dealing with concept drift (statistical properties of modelled time series change over time)
+     * For instance originally monotonic time series changed to trend stationary or seasonal pattern showed up.
+     */
+    public abstract static class ConceptDriftStrategy {
+
+        protected AutomaticForecaster forecaster;
+
+        /**
+         * @param learnSize size of dataSet from learn method
+         * @return true if model should be selected - concept drift is present
+         */
+        public abstract boolean recalculateModel(int learnSize);
+    }
+
+    /**
+     * Periodically after x learned points selects the best model.
+     */
+    public static class PeriodicIntervalStrategy extends ConceptDriftStrategy {
+
+        private final int period;
+
+        /**
+         * @param period number of learning points after which the best model is selected
+         */
+        public PeriodicIntervalStrategy(int period) {
+            if (period < 1) {
+                throw new IllegalArgumentException("Period should be > 1");
+            }
+
+            this.period = period;
+        }
+
+        @Override
+        public boolean recalculateModel(int learnSize) {
+            return (forecaster.counter + learnSize) >= period;
+        }
+
+        public int getPeriod() {
+            return period;
+        }
+    }
+
+    /**
+     * When the best model is selected initial accuracy statistics are calculated (MSE, MAE).
+     * On following learning the same accuracy statistics are calculated. This strategy compares these two
+     * statistics and if changed more by x percent best model selection is triggered.
+     */
+    public static class ErrorChangeStrategy extends ConceptDriftStrategy {
+
+        private final int percentageChange;
+        private final Statistics statistics;
+
+        public ErrorChangeStrategy(int percentageChange, Statistics statistics) {
+            if (percentageChange > 100 || percentageChange < 1) {
+                throw new IllegalArgumentException("Change should be between 1-100");
+            }
+
+            this.percentageChange = percentageChange;
+            this.statistics = statistics;
+        }
+
+        @Override
+        public boolean recalculateModel(int learnSize) {
+            double initStat = statistics(forecaster.model().initStatistics(), statistics);
+            double runStat = statistics(forecaster.model().runStatistics(), statistics);
```

Comment:
Do you mean store concrete statistics (MSE, MAE) in `ErrorChangeStrategy`?

Would it be much faster? it would avoid:
1. `forecaster.model().initStatistics()`
2. `statistics(forecaster.model().initStatistics(), statistics)`
3. `if (statistics == Statistics.MAE)`
= two method calls and one if